src/main/java/org/radarcns/phone/PhoneBluetoothManager.java

```diff
@@ -0,0 +1,106 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.radarcns.phone;
+
+import android.bluetooth.BluetoothAdapter;
+import android.bluetooth.BluetoothDevice;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.support.annotation.NonNull;
+import org.radarcns.android.data.DataCache;
+import org.radarcns.android.device.AbstractDeviceManager;
+import org.radarcns.android.device.BaseDeviceState;
+import org.radarcns.android.device.DeviceStatusListener;
+import org.radarcns.key.MeasurementKey;
+
+import java.util.Set;
+
+public class PhoneBluetoothManager extends AbstractDeviceManager<PhoneBluetoothService, BaseDeviceState> implements Runnable {
+    private static final int SCAN_DEVICES_REQUEST_CODE = 3248902;
+    private static final String ACTION_SCAN_DEVICES = "org.radarcns.phone.PhoneBluetoothManager.ACTION_SCAN_DEVICES";
+    private final OfflineProcessor processor;
+    private final DataCache<MeasurementKey, PhoneBluetoothDevices> bluetoothDevicesTable;
+
+    public PhoneBluetoothManager(PhoneBluetoothService service) {
+        super(service, service.getDefaultState(), service.getDataHandler(), service.getUserId(), service.getSourceId());
+
+        processor = new OfflineProcessor(service, this, SCAN_DEVICES_REQUEST_CODE,
+                ACTION_SCAN_DEVICES, service.getCheckInterval(), true);
+
+        bluetoothDevicesTable =  getCache(service.getTopics().getBluetoothDevicesTopic());
+    }
+
+
+    @Override
+    public void run() {
+        final BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
+        final boolean wasEnabled = bluetoothAdapter.isEnabled();
+
+        if (!wasEnabled) {
+            bluetoothAdapter.enable();
+        }
+
+        IntentFilter filter = new IntentFilter();
+        filter.addAction(BluetoothDevice.ACTION_FOUND);
+        filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED);
+
+        BroadcastReceiver bluetoothReceiver = new BroadcastReceiver() {
+            private int numberOfDevices;
+
+            @Override
+            public void onReceive(Context context, Intent intent) {
+                switch (intent.getAction()) {
+                    case BluetoothDevice.ACTION_FOUND: {
+                        numberOfDevices++;
+                        break;
+                    }
+
+                    case BluetoothAdapter.ACTION_DISCOVERY_FINISHED: {
+                        getService().unregisterReceiver(this);
```

Comment:
One more question: is it possible that the service gets reset (`close()`) after startDiscovery has been called and before `unregisterReceiver` has been called? Could this result in some invalid state here, when the service has already been shut down? If yes, perhaps the bluetoothReceiver could be a field created at `start(Set)`, registered at `run()`, deregistered at `ACTION_DISCOVERY_FINISHED` and also deregistered at `close()`. The deregistration could happen in a synchronised block, where it is deregistered and set to `null`, only if it is non-`null`.

Reply:
I don't think that it could result in any invalid state but there was a more important issue: the close method wasn't overridden at all, so the processor never stopped.

Reply:
Workflow:
1. `processor` is started and `run()` is called, and the `BroadcastReceiver` is registered
2. the `DeviceService` gets stopped
3. `close()` gets called and the processor is stopped
4. the service is shutdown

However, the broadcast listener is still registered. Android will throw a warning/error like "BroadcastReceiver was leaked". The current broadcast receiver is registered to a service that is no longer running. This can be solved by also removing this receiver in `close()`.

Reply:
Fixed

Reply:
One more question: could you set `bluetoothBroadcastReceiver` to null here as well?

Reply:
Done